src/storage-client/src/types/sources.rs

```diff
@@ -2759,13 +2762,79 @@ impl Codec for SourceData {
     }
 }
 
+/// An implementation of [PartEncoder] for [SourceData].
+///
+/// This mostly delegates the encoding logic to [RowEncoder], but flatmaps in
+/// an Err column.
+#[derive(Debug)]
+pub struct SourceDataEncoder<'a> {
+    wrapped: RowEncoder<'a>,
+}
+
+impl<'a> PartEncoder<'a, SourceData> for SourceDataEncoder<'a> {
+    fn encode(&mut self, val: &SourceData) {
+        match val.as_ref() {
+            Ok(row) => {
+                // Append a fake Err datum.
+                let ok_datums = row.iter();
+                let err_datum = Datum::Null;
+                let datums = ok_datums.chain(std::iter::once(err_datum));
+                for (encoder, datum) in self.wrapped.col_encoders().iter_mut().zip(datums) {
+                    encoder.encode(datum);
+                }
+            }
+            Err(err) => {
+                let err = err.into_proto().encode_to_vec();
+
+                // Prepend a fake datum for every Ok column.
+                let ok_datums =
+                    std::iter::repeat(Datum::Null).take(self.wrapped.col_encoders().len());
+                let err_datum = Datum::Bytes(&err);
+                let datums = ok_datums.chain(std::iter::once(err_datum));
+                for (encoder, datum) in self.wrapped.col_encoders().iter_mut().zip(datums) {
+                    encoder.encode(datum);
+                }
+            }
+        }
+    }
+}
+
+const SOURCE_DATA_ERROR: &str = "mz_internal_super_secret_source_data_errors";
```

Comment:
would love to get rid of this hack... maybe the answer is just pushing the union type into parquet

Reply:
yeah, agreed that it feels bad. the union type will be pretty complicated though, so seems like followup work to me?

Reply:
I'm fine with that, agreed the union type will be tricky